File: ItalianCannon/Configurations.vb
Public Class Configurations
    Public Class ConfObj
        Public Note As String
        Public TeaCupTargets As List(Of AtkUrl)
        Public IntervalPerThread As Integer
        Public MaxRequestsPerThread As Long
        Public UserAgent As String
        Public AppearsToBeDefault As Boolean
        Public DisableSSLValidation As Boolean
        Public IgnoreHTTPError As Boolean
        Public ExtraHTTPHeaders As List(Of Header)
        Public EnableAnimations As Boolean
        Public EnableColors As Boolean
        Public SilentMode As Boolean
        Public RandomPOST As Boolean
        Public RandomPOSTRate As Byte
        Public RandomPOSTLength As Long
        Public RandomHEAD As Boolean
        Public RandomHEADRate As Byte
        Public RandomQuery As Boolean
        Sub New()
            Note = "Please change 'AppearsToBeDefault' to False after changing settings. ItalianCannon will ignore this configuration entry. For headers help, see https://see.wtf/O5vFp"
            IntervalPerThread = 500
            MaxRequestsPerThread = 1000
            UserAgent = "Mozilla/5.0 (Linux) AppleWebKit/888.88 (KHTML, like Gecko) Chrome/66.6.2333.66 Safari/233.33"
            AppearsToBeDefault = True
            DisableSSLValidation = False
            IgnoreHTTPError = False
            Dim SampleTCTCol As New List(Of AtkUrl)
            SampleTCTCol.Add(New AtkUrl)
            TeaCupTargets = SampleTCTCol
            Dim SampleHeaderCol As New List(Of Header)
            SampleHeaderCol.Add(New Header)
            ExtraHTTPHeaders = SampleHeaderCol
            EnableAnimations = False
            EnableColors = True
            SilentMode = False
            RandomPOST = False
            RandomPOSTRate = 1
            RandomPOSTLength = 128
            RandomHEAD = False
            RandomHEADRate = 1
            RandomQuery = False
        End Sub
    End Class

    Public Class Header
        Public HType As System.Net.HttpRequestHeader
        Public Content As String
        Sub New()
            HType = System.Net.HttpRequestHeader.Allow
            Content = "GET"
        End Sub
    End Class

    Public Class AtkUrl
        Public Url As String
        Public Threads As Integer
        Sub New()
            Url = "https://www.contoso.com"
            Threads = 1
        End Sub
    End Class

    Public Shared Sub Initiate()
        'Use default conf for failsafe.
        Constants.CurrentConfigurations = New ConfObj()
        'Find file, if does not exist, create one.
        Out("Loading configurations...", "CONF")
        If Not IO.File.Exists(Constants.ConfFile) Then
            Out("Configurations file not found. Creating a new one at " & Constants.ConfFile & ".", "CONF", LogLevels.WARN,, True)
            Constants.CurrentConfigurations = New ConfObj
            SaveConf()
            WaitEdit() 'Display message and exit.
        Else
            Try
                ReadConf()
                If Constants.CurrentConfigurations.AppearsToBeDefault Then
                    Out("Configurations appear to be default.", "CONF", LogLevels.WARN,, True)
                    WaitEdit()
                    Exit Sub 'Since the waitedit sub itself will output a "Complete", so I just let it "go".
                End If
                If Not CheckConflictConf() Then
                    Out("Conflict options detected.", "CONF", LogLevels.EXCEPTION,, True)
                    Environment.Exit(1)
                End If
                Out("Complete.", "CONF")
            Catch ex As Exception
                Out("Cannot read configuration: " & ex.ToString, "CONF", LogLevels.EXCEPTION,, True)
                Environment.Exit(1)
            End Try
        End If
    End Sub

    Public Shared Sub SaveConf()
        Constants.CurrentConfigurations = Constants.CurrentConfigurations
        Dim text As String = Newtonsoft.Json.JsonConvert.SerializeObject(Constants.CurrentConfigurations, Newtonsoft.Json.Formatting.Indented)
        Dim writer As New IO.StreamWriter(IO.File.Create(Constants.ConfFile), System.Text.Encoding.UTF8)
        writer.Write(text)
        writer.Flush()
        writer.Close()
    End Sub

    Public Shared Sub ReadConf()
        Dim settingsText As String = IO.File.ReadAllText(Constants.ConfFile)
        Constants.CurrentConfigurations = Newtonsoft.Json.JsonConvert.DeserializeObject(Of ConfObj)(settingsText)
        With Constants.CurrentConfigurations
            .TeaCupTargets.RemoveAt(0)
            .ExtraHTTPHeaders.RemoveAt(0)
        End With
    End Sub

    Public Shared Sub WaitEdit()
        Out("Please change the configurations and try again.", "CONF", LogLevels.WARN,, True)
        Environment.Exit(0)
    End Sub

    Public Shared Function CheckConflictConf() As Boolean
        If (Constants.CurrentConfigurations.SilentMode And Constants.CurrentConfigurations.EnableAnimations) Then Return False
        Return True
    End Function
End Class

File: src/Configurations.cs
// ItalianCannon.Configurations
using ItalianCannon;
using Microsoft.VisualBasic.CompilerServices;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Text;

namespace ItalianCannon
{
    public class Configurations
    {
        public class ConfObj
        {
            public string Note;

            public List<AtkUrl> TeaCupTargets;

            public int IntervalPerThread;

            public long MaxRequestsPerThread;

            public string UserAgent;

            public bool AppearsToBeDefault;

            public bool DisableSSLValidation;

            public bool IgnoreHTTPError;

            public List<Header> ExtraHTTPHeaders;

            public bool EnableAnimations;

            public bool EnableColors;

            public bool VerboseMode;

            public bool RandomPOST;

            public byte RandomPOSTRate;

            public long RandomPOSTLength;

            public bool RandomHEAD;

            public byte RandomHEADRate;

            public ConfObj()
            {
                this.Note = "Please change 'AppearsToBeDefault' to False after changing settings. ItalianCannon will ignore this configuration entry. For headers help, see https://github.com/dotnet/corefx/blob/master/src/System.Net.WebHeaderCollection/src/System/Net/HttpRequestHeader.cs";
                this.IntervalPerThread = 500;
                this.MaxRequestsPerThread = 1000L;
                this.UserAgent = "Mozilla/5.0 (Linux) AppleWebKit/888.88 (KHTML, like Gecko) Chrome/66.6.2333.66 Safari/233.33";
                this.AppearsToBeDefault = true;
                this.DisableSSLValidation = false;
                this.IgnoreHTTPError = false;
                this.TeaCupTargets = new List<AtkUrl>
                {
                    new AtkUrl()
                };
                this.ExtraHTTPHeaders = new List<Header>
                {
                    new Header()
                };
                this.EnableAnimations = false;
                this.EnableColors = true;
                this.VerboseMode = false;
                this.RandomPOST = false;
                this.RandomPOSTRate = 1;
                this.RandomPOSTLength = 128L;
                this.RandomHEAD = false;
                this.RandomHEADRate = 1;
            }
        }

        public class Header
        {
            public HttpRequestHeader HType;

            public string Content;

            public Header()
            {
                this.HType = HttpRequestHeader.Allow;
                this.Content = "GET";
            }
        }

        public class AtkUrl
        {
            public string Url;

            public int Threads;

            public AtkUrl()
            {
                this.Url = "https://www.baidu.com";
                this.Threads = 1;
            }
        }

        public static void Initiate()
        {
            Constants.CurrentConfigurations = new ConfObj();
            SOPT.Out("Loading configurations...", "CONF", SOPT.LogLevels.INFO, true, false);
            if (!File.Exists(Constants.ConfFile))
            {
                SOPT.Out("Configurations file not found. Creating a new one at " + Constants.ConfFile + ".", "CONF", SOPT.LogLevels.WARN, true, true);
                Constants.CurrentConfigurations = new ConfObj();
                Configurations.SaveConf();
                Configurations.WaitEdit();
            }
            else
            {
                try
                {
                    Configurations.ReadConf();
                    if (Constants.CurrentConfigurations.AppearsToBeDefault)
                    {
                        SOPT.Out("Configurations appear to be default.", "CONF", SOPT.LogLevels.WARN, true, true);
                        Configurations.WaitEdit();
                    }
                    else
                    {
                        if (!Configurations.CheckConflictConf())
                        {
                            SOPT.Out("Conflict options detected.", "CONF", SOPT.LogLevels.EXCEPTION, true, true);
                            Environment.Exit(1);
                        }
                        SOPT.Out("Complete.", "CONF", SOPT.LogLevels.INFO, true, false);
                    }
                }
                catch (Exception ex2)
                {
                    ProjectData.SetProjectError(ex2);
                    Exception ex = ex2;
                    SOPT.Out("Cannot read configuration: " + ex.ToString(), "CONF", SOPT.LogLevels.EXCEPTION, true, true);
                    Environment.Exit(1);
                    ProjectData.ClearProjectError();
                }
            }
        }

        public static void SaveConf()
        {
            Constants.CurrentConfigurations = Constants.CurrentConfigurations;
            string text = JsonConvert.SerializeObject(Constants.CurrentConfigurations, Formatting.Indented);
            StreamWriter streamWriter = new StreamWriter(File.Create(Constants.ConfFile), Encoding.UTF8);
            streamWriter.Write(text);
            streamWriter.Flush();
            streamWriter.Close();
        }

        public static void ReadConf()
        {
            Constants.CurrentConfigurations = JsonConvert.DeserializeObject<ConfObj>(File.ReadAllText(Constants.ConfFile));
            ConfObj currentConfigurations = Constants.CurrentConfigurations;
            currentConfigurations.TeaCupTargets.RemoveAt(0);
            currentConfigurations.ExtraHTTPHeaders.RemoveAt(0);
        }

        public static void WaitEdit()
        {
            SOPT.Out("Please change the configurations and try again.", "CONF", SOPT.LogLevels.WARN, true, true);
            Environment.Exit(0);
        }

        public static bool CheckConflictConf()
        {
            if (Constants.CurrentConfigurations.VerboseMode & Constants.CurrentConfigurations.EnableAnimations)
            {
                return false;
            }
            return true;
        }
    }
}